### PR TITLE
Retrieve log entries from the status server.

### PR DESCRIPTION
**Pull request by BramGruneir**
This enables the status server to retrieve log entires directly from the log files.

To get a list of logs, just go to: 
  **_status/local/log/**
or if you like to choose your logging level 
  **_status/local/log/WARNING**
or if you want to choose your start and end time
  **_status/local/log/ERROR?starttime=1433884484989655088&endtime=1433884484989655088**

But don't forget to turn on file logging using the flags:
--log-dir=/tmp/logs --alsologtostderr=true --logtostderr=false

### CODE REVIEW

File: util/log/log.go [hunk 1]
```diff
@@ -17,7 +17,52 @@
...
+import (
+	"strings"
+
+	"golang.org/x/net/context"
+)
+
+// Level Enum for log severity levels.
+type Level int
+
+// TODO(Bram): combine this with clog.severity, and rename severity.
```
Comment (BramGruneir) [inline util/log/log.go]: I'll have a new PR out to merge this with clog.severity.  I wrote most of this and didn't know it existed.

File: util/log/file.go [hunk 2]
```diff
@@ -221,3 +296,92 @@ func GetLogReader(filename string, allowAbsolute bool) (io.ReadCloser, error) {
...
 }
+
+// int64sortable is required so we can sort on an int64 slice.
+type int64sortable []int64
+
+func (a int64sortable) Len() int           { return len(a) }
+func (a int64sortable) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
+func (a int64sortable) Less(i, j int) bool { return a[i] < a[j] }
+
+// FetchEntiresFromFiles fetches all available logs on disk that are of the
```
Comment (mrtracy) [inline util/log/file.go]: s/Entires/Entries/g ?

File: util/log/file.go [hunk 3]
```diff
@@ -221,3 +296,92 @@ func GetLogReader(filename string, allowAbsolute bool) (io.ReadCloser, error) {
...
+// will be retrieved. The logs entries returned will be in decreasing order,
+// with the closest log to the start time being the first entry.
+func FetchEntiresFromFiles(level Level, startTimeNano, endTimeNano int64) ([]proto.LogEntry, error) {
+	logFiles, err := ListLogFiles()
+	if err != nil {
+		return nil, err
+	}
+
+	logFileMap := make(map[int64]FileInfo)
+	var nanos int64sortable
```
Comment (mrtracy) [inline util/log/file.go]: Weird variable name, its really timestamps expressed as nanos.

File: util/log/file.go [hunk 4]
```diff
@@ -221,3 +296,92 @@ func GetLogReader(filename string, allowAbsolute bool) (io.ReadCloser, error) {
...
+	var diffToStart int64 = math.MaxInt64
+	// closestToStartNano holds the last log file that would include any entries
+	// from the start time onward in it.
+	var closestToStartNano int64
+	for _, logFile := range logFiles {
+		if logFile.Details.Level == level {
```
Comment (mrtracy) [inline util/log/file.go]: This is just a preference, but in cases like to use `continue` statements to avoid excessive indentation.

```
if logFile.Details.Level != level {
   continue;
}
```

File: util/log/log.go (hunk 1 above)
Comment (tamird) [inline util/log/log.go]: Why not fold it into this PR?

File: util/log/log.go [hunk 5]
```diff
@@ -17,7 +17,52 @@
...
+// These constants identify the log levels in order of increasing severity.
+const (
+	INFO Level = iota
+	WARNING
+	ERROR
+	FATAL
+)
+
+var levels = [...]string{
+	"INFO",
```
Comment (tamird) [inline util/log/log.go]: this should use stringer

I think you should merge this with clog.severity here so we can avoid reviewing both

File: server/status_test.go [hunk 6]
```diff
@@ -337,20 +342,65 @@ func TestStatusLocalLog(t *testing.T) {
...
 			}
 		}
 		if !found {
-			t.Errorf("exected to find test message in %v", logW.Data)
+			t.Errorf("expected to find test message in %v", logW.Data)
+		}
+	}
+
+	// Fetch the full list of log entries.
+	type entryWarpper struct {
```
Comment (mrtracy) [inline server/status_test.go]: s/Warpper/Wrapper/g

File: util/log/file.go [hunk 7]
```diff
@@ -46,7 +56,8 @@ var logDir *string
 var logDirs []string
 
 // logFileRE matches log files to avoid exposing non-log files accidentally.
-var logFileRE = regexp.MustCompile(`log\.(INFO|WARNING|ERROR)\.`)
+// and it splits the details of the filename into groups for easy parsing.
+var logFileRE = regexp.MustCompile(`([^\.]+)\.([^\.]+)\.([^\.]+)\.log\.(ERROR|WARNING|INFO)\.([^\.]+)\.(\d+)`)
```
Comment (mrtracy) [inline util/log/file.go]: Should include an example of the format in this comment, or in the parse method below.

File: server/status.go [hunk 8]
```diff
@@ -51,13 +53,22 @@ const (
...
 
 	// statusLocalLogKeyPrefix exposes a list of log files for the node.
-	// logs -> lists available log files
-	// logs/ -> lists available log files
-	// logs/{file} -> fetches contents of named log
-	statusLocalLogKeyPrefix = statusLocalKeyPrefix + "logs/"
+	// logfiles -> lists available log files
+	// logfiles/ -> lists available log files
+	// logfiles/{file} -> fetches contents of named log
+	statusLocalLogFileKeyPrefix = statusLocalKeyPrefix + "logfiles/"
```
Comment (mrtracy) [inline server/status.go]: This doesn't match your commit message - why the change from 'logs'?

File: server/status.go [hunk 9]
```diff
@@ -223,6 +236,66 @@ func (s *statusServer) handleLocalLog(w http.ResponseWriter, r *http.Request, ps
...
+		if err != nil {
+			log.Error(err)
+			w.WriteHeader(http.StatusInternalServerError)
+			return
+		}
+	} else {
+		startTimeNano = time.Now().AddDate(0, 0, -1).UnixNano()
+	}
+
+	endTime := r.URL.Query().Get("endtime")
```
Comment (mrtracy) [inline server/status.go]: These look like they could be extracted into a common method; a parseint with default.

File: util/log/file.go [hunk 10]
```diff
@@ -221,3 +296,92 @@ func GetLogReader(filename string, allowAbsolute bool) (io.ReadCloser, error) {
...
+		newEntries, err := readAllEntriesFromFile(logFileMap[nano], startTimeNano, endTimeNano)
+		if err != nil {
+			return nil, err
+		}
+		entries = append(entries, newEntries...)
+		if len(entries) >= EntiresCutoff {
+			break
+		}
+		if nano == closestToStartNano {
+			// don't read any files that have no timestamps after
```
Comment (mrtracy) [inline util/log/file.go]: Comment has poor grammar.

File: util/log/file.go [hunk 11]
```diff
@@ -221,3 +296,92 @@ func GetLogReader(filename string, allowAbsolute bool) (io.ReadCloser, error) {
...
+	logFileMap := make(map[int64]FileInfo)
+	var nanos int64sortable
+	var diffToStart int64 = math.MaxInt64
+	// closestToStartNano holds the last log file that would include any entries
+	// from the start time onward in it.
+	var closestToStartNano int64
+	for _, logFile := range logFiles {
+		if logFile.Details.Level == level {
+			nano := logFile.Details.Time.UnixNano()
+			if nano <= endTimeNano {
```
Comment (mrtracy) [inline util/log/file.go]: Really don't like the use of "nano" here, this is a log timestamp.

File: util/log/file.go [hunk 12]
```diff
@@ -221,3 +296,92 @@ func GetLogReader(filename string, allowAbsolute bool) (io.ReadCloser, error) {
...
+				logFileMap[nano] = logFile
+				nanos = append(nanos, nano)
+				if nano <= startTimeNano && (startTimeNano-nano) < diffToStart {
```
Comment (mrtracy) [inline util/log/file.go]: I don't think this provides any sort of optimization.  If another log was already encountered with a smaller diffToStart, then there's no need to add the log to the collection.

If you're fine with adding every log with (time <= endTimeNano) to the slice, as it is currently doing, then you don't need to track closestToStartNano - in the sorted processing loop below, you could just bail after processing the first log with a timestamp less than startTime.

File: util/log/file.go [hunk 13]
```diff
@@ -221,3 +296,92 @@ func GetLogReader(filename string, allowAbsolute bool) (io.ReadCloser, error) {
...
+		}
+	}
+
+	// There are no logs to display.
+	if len(nanos) == 0 {
+		return []proto.LogEntry{}, nil
+	}
+
+	// Sort the files in reverse order so we will fetch the newest first.
+	sort.Sort(sort.Reverse(nanos))
```
Comment (mrtracy) [inline util/log/file.go]: It doesn't look like there's any test for all of this sorting and combination logic.  I think it could be tested pretty easily if it was separated from the file system operations.

File: util/log/log.go (hunk 1 above)
Comment (BramGruneir) [inline util/log/log.go]: I already was working on it.  Done.

File: util/log/file.go (hunk 2 above)
Comment (BramGruneir) [inline util/log/file.go]: fixed

File: util/log/file.go (hunk 3 above)
Comment (BramGruneir) [inline util/log/file.go]: Replaced nanos and nano with timestamps and timestamp respectively.

File: util/log/log.go (hunk 5 above)
Comment (BramGruneir) [inline util/log/log.go]: this has been merged into with clog.severity.  It already had a String() function that returns an int, so I created Name().

File: server/status_test.go (hunk 6 above)
Comment (BramGruneir) [inline server/status_test.go]: fixed, but warper sounded so much cooler.

File: util/log/file.go (hunk 7 above)
Comment (BramGruneir) [inline util/log/file.go]: Done.

File: server/status.go (hunk 9 above)
Comment (BramGruneir) [inline server/status.go]: Done.  Not sure how much this saves, as it still needs to do the same error checking.  But it goes look a bit nicer.

File: util/log/file.go (hunk 11 above)
Comment (BramGruneir) [inline util/log/file.go]: Agreed, fixed.

File: util/log/file.go (hunk 10 above)
Comment (BramGruneir) [inline util/log/file.go]: fixed

File: util/log/file.go (hunk 4 above)
Comment (BramGruneir) [inline util/log/file.go]: Done.  I usually gauge this based on the depth of the inner statement, which grew after I originally wrote it.

File: util/log/file.go (hunk 12 above)
Comment (BramGruneir) [inline util/log/file.go]: This is a great idea.  I implemented a flag in read all entries from file that returns true if any entry had a timestamp less than startTime.

File: util/log/file.go (hunk 13 above)
Comment (BramGruneir) [inline util/log/file.go]: Good call.  Pulling it out for easier testing.

File: server/status.go [hunk 14]
```diff
@@ -223,6 +236,72 @@ func (s *statusServer) handleLocalLog(w http.ResponseWriter, r *http.Request, ps
...
+
+// handleLocalLog returns the log entires parsed from the log file that occurred
+// after or on the query parameter "starttime" and before or on the query
+// parameter "endtime". Additionally, if a level is specified via the URL, the
+// log entires are scoped to that level. If no "starttime" is provided, then
+// a "starttime" of a day ago is used.  If no "endtime" is provided, the current
+// time is used. The cutoff for entires is defined in log.EntriesCutoff.
+func (s *statusServer) handleLocalLog(w http.ResponseWriter, r *http.Request, ps httprouter.Params) {
+	log.Flush()
+	level, _ := log.SeverityByName(ps.ByName("level"))
```
Comment (bdarnell) [inline server/status.go]: This should return an error on an unrecognized level (just like we return errors for unparseable numbers).

File: util/log/file.go [hunk 15]
```diff
@@ -28,16 +30,23 @@ import (
...
+	"github.com/cockroachdb/cockroach/proto"
 	"github.com/cockroachdb/cockroach/util"
 )
 
 // MaxSize is the maximum size of a log file in bytes.
 var MaxSize uint64 = 1024 * 1024 * 1800
 
+// EntiresCutoff is the cutoff in which FetchEntiresFromFiles will stop
+// reading in older log files.
+var EntiresCutoff = 100000
```
Comment (bdarnell) [inline util/log/file.go]: s/Entires/Entries/g

File: util/log/file.go [hunk 16]
```diff
@@ -221,3 +301,103 @@ func GetLogReader(filename string, allowAbsolute bool) (io.ReadCloser, error) {
...
+	}
+	return result
+}
+
+// FetchEntriesFromFiles fetches all available logs on disk that are of the
+// level of severity (or worse) and are between the start and end times. It will
+// stop reading in new files if the EntiresCutoff is exceeded and no more files
+// will be retrieved. The logs entries returned will be in decreasing order,
+// with the closest log to the start time being the first entry.
+func FetchEntriesFromFiles(severity Severity, startTimestamp, endTimestamp int64) ([]proto.LogEntry, error) {
```
Comment (bdarnell) [inline util/log/file.go]: `EntriesCutoff` should be a parameter to this function, so we can have a reasonable default but exceed it on demand.

File: util/log/file.go [hunk 17]
```diff
@@ -221,3 +301,103 @@ func GetLogReader(filename string, allowAbsolute bool) (io.ReadCloser, error) {
...
+func readAllEntriesFromFile(file FileInfo, startTimestamp, endTimestamp int64) ([]proto.LogEntry, bool, error) {
+	reader, err := GetLogReader(file.Name, false)
+	defer reader.Close()
+	if reader == nil || err != nil {
+		return nil, false, err
+	}
+	entries := []proto.LogEntry{}
+	decoder := NewEntryDecoder(reader)
+	entryBeforeStart := false
+	for {
```
Comment (bdarnell) [inline util/log/file.go]: EntriesCutoff needs to be handled in this loop as well; we don't want to load a whole log file into memory just to discard most of it.

File: util/log/file.go (hunk 16 above)
Comment (BramGruneir) [inline util/log/file.go]: Sure, I can add that.

File: util/log/file.go (hunk 17 above)
Comment (BramGruneir) [inline util/log/file.go]: Right now, the cutoff just directs it not to process any more files if exceeded.  I didn't make it a hard limit.  But I can.

File: util/log/file.go (hunk 15 above)
Comment (BramGruneir) [inline util/log/file.go]: Bah,  I mistyped it everywhere.  Fixed.

File: server/status.go (hunk 14 above)
Comment (BramGruneir) [inline server/status.go]: done.

File: util/log/file.go (hunk 17 above)
Comment (bdarnell) [inline util/log/file.go]: It should be a hard limit. Log files are allowed to grow to gigabytes (MaxSize=1.8GB).

Comment (BramGruneir) [inline util/log/file.go]: Yeah, I just looked at that.  Why such a high max?  At MS, we limited log files to either 1 or 10 megs each.  So you could easily grab the file when needed and it could be loaded in most editors.

Comment (bdarnell) [inline util/log/file.go]: I'm not sure why it was so high. Maybe there were issues at google with large numbers of files in one directory (google's initial log aggregation system was to export the log directory from each machine over nfs). Lowering the limit to something that can be held in memory seems like a good idea to me.

Comment (BramGruneir) [inline util/log/file.go]: I've set the max at 10 megs for now. And implemented the hard limit, with the ability to override it.

**Comment by BramGruneir** [comment]
PTAL, all comments addressed, and there are no more instances of entries spelt as entires.

**Comment by mrtracy** [comment]
LGTM

**Comment by BramGruneir** [comment]
Oh, I plan to.  Just wanted to leave them as is for reviewing purposes.

File: server/status.go [hunk 18]
```diff
@@ -223,6 +239,99 @@ func (s *statusServer) handleLocalLog(w http.ResponseWriter, r *http.Request, ps
...
+	level := ps.ByName("level")
+	var sev log.Severity
+	if len(level) == 0 {
+		sev = log.InfoLog
+	} else {
+		var sevFound bool
+		sev, sevFound = log.SeverityByName(level)
+		if !sevFound {
+			log.Errorf("could not determine severity %s", level)
+			w.WriteHeader(http.StatusInternalServerError)
```
Comment (tamird) [inline server/status.go]: this should be a bad request, not internal error

File: server/status.go [hunk 19]
```diff
@@ -223,6 +239,99 @@ func (s *statusServer) handleLocalLog(w http.ResponseWriter, r *http.Request, ps
...
+			return
+		}
+	}
+
+	startTimestamp, err := parseInt64WithDefault(
+		r.URL.Query().Get("starttime"),
+		time.Now().AddDate(0, 0, -1).UnixNano())
+	if err != nil {
+		log.Error(err)
+		w.WriteHeader(http.StatusInternalServerError)
```
Comment (tamird) [inline server/status.go]: bad request

File: server/status.go [hunk 20]
```diff
@@ -223,6 +239,99 @@ func (s *statusServer) handleLocalLog(w http.ResponseWriter, r *http.Request, ps
...
+		return
+	}
+
+	endTimestamp, err := parseInt64WithDefault(
+		r.URL.Query().Get("endtime"),
+		time.Now().UnixNano())
+	if err != nil {
+		log.Error(err)
+		w.WriteHeader(http.StatusInternalServerError)
```
Comment (tamird) [inline server/status.go]: :recycle:

File: server/status.go (hunk 19 above)
Comment (tamird) [inline server/status.go]: all the errors here should be reported back to the client rather than logged

File: util/log/file.go [hunk 21]
```diff
@@ -46,7 +51,14 @@ var logDir *string
 var logDirs []string
 
 // logFileRE matches log files to avoid exposing non-log files accidentally.
-var logFileRE = regexp.MustCompile(`log\.(INFO|WARNING|ERROR)\.`)
+// and it splits the details of the filename into groups for easy parsing.
```
Comment (tamird) [inline util/log/file.go]: the previous sentence ended with a period

File: util/log/file.go [hunk 22]
```diff
@@ -85,36 +97,92 @@ func shortHostname(hostname string) string {
...
-		tag,
-		t.Year(),
-		t.Month(),
-		t.Day(),
-		t.Hour(),
-		t.Minute(),
-		t.Second(),
+// escapeString replaces all the periods in the string with an underscore, and
+// every underscore with a double underscore.
+func escapeStringForFilename(s string) string {
```
Comment (tamird) [inline util/log/file.go]: can you add a comment indicating why this is needed?

File: server/status.go [hunk 23]
```diff
@@ -223,6 +239,99 @@ func (s *statusServer) handleLocalLog(w http.ResponseWriter, r *http.Request, ps
...
+	maxEntries, err := parseInt64WithDefault(
+		r.URL.Query().Get("max"),
+		defaultMaxLogEntries)
+	if err != nil {
+		log.Error(err)
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+	if maxEntries < 1 {
+		log.Errorf("max must be set to a value greater than 1")
```
Comment (tschottdorf) [inline server/status.go]: s/1/0/

File: util/log/file.go [hunk 24]
```diff
@@ -85,36 +97,92 @@ func shortHostname(hostname string) string {
...
+}
+
+func parseLogFilename(filename string) (FileDetails, error) {
+	matches := logFileRE.FindStringSubmatch(filename)
+	if matches == nil || len(matches) != 7 {
+		return FileDetails{}, util.Errorf("not a log file")
+	}
+
+	sev, sevFound := SeverityByName(matches[4])
+	if !sevFound {
```
Comment (tamird) [inline util/log/file.go]: under what circumstance can this happen?

File: util/log/file.go [hunk 25]
```diff
@@ -85,36 +97,92 @@ func shortHostname(hostname string) string {
...
+	UserName string
+	Severity Severity
+	Time     time.Time
+	PID      int
+}
+
+func parseLogFilename(filename string) (FileDetails, error) {
+	matches := logFileRE.FindStringSubmatch(filename)
+	if matches == nil || len(matches) != 7 {
+		return FileDetails{}, util.Errorf("not a log file")
```
Comment (tamird) [inline util/log/file.go]: "malformed log filename"?

File: util/log/file.go [hunk 26]
```diff
@@ -85,36 +97,92 @@ func shortHostname(hostname string) string {
...
+		return FileDetails{}, err
+	}
+
+	return FileDetails{
+		Program:  unescapeStringForFilename(matches[1]),
+		Host:     unescapeStringForFilename(matches[2]),
+		UserName: unescapeStringForFilename(matches[3]),
+		Severity: sev,
+		Time:     time,
+		PID:      int(pid),
```
Comment (tamird) [inline util/log/file.go]: perhaps OT: shouldn't this be an unsigned type?

File: util/log/clog.go [hunk 27]
```diff
@@ -45,66 +45,73 @@ import (
...
 
 const severityChar = "IWEF"
 
 // SeverityName provides a mapping from Severity level to a string.
-var severityName = []string{
-	infoLog:    "INFO",
-	warningLog: "WARNING",
-	errorLog:   "ERROR",
-	fatalLog:   "FATAL",
+var SeverityName = []string{
```
Comment (tschottdorf) [inline util/log/clog.go]: why does this need to be exported? Also, why not use Stringer (doesn't have to be in this PR)

File: server/status.go (hunk 18 above)
Comment (BramGruneir) [inline server/status.go]: fixed.

File: server/status.go (hunk 19 above)
Comment (BramGruneir) [inline server/status.go]: fixed

File: server/status.go (hunk 20 above)
Comment (BramGruneir) [inline server/status.go]: done.

File: util/log/file.go [hunk 28]
```diff
@@ -139,14 +207,18 @@ func create(tag string, t time.Time) (f *os.File, filename string, err error) {
 
 // verifyFileInfo verifies that the file specified by filename is a
 // regular file and filename matches the expected filename pattern.
-// Returns nil on success; otherwise error.
-func verifyFileInfo(info os.FileInfo) error {
+// Returns the log file details success; otherwise error.
+func verifyFileInfo(info os.FileInfo) (FileDetails, error) {
```
Comment (tamird) [inline util/log/file.go]: this function needs a rename

File: util/log/file.go (hunk 24 above)
Comment (BramGruneir) [inline util/log/file.go]: Actually, not a log file is used to weed out the files that are not log files.

File: util/log/file.go [hunk 29]
```diff
@@ -221,3 +297,111 @@ func GetLogReader(filename string, allowAbsolute bool) (io.ReadCloser, error) {
 	}
 	return nil, err
 }
+
+// int64sortable is required so we can sort on an int64 slice.
+type int64sortable []int64
```
Comment (tamird) [inline util/log/file.go]: why not make this a type alias of `[]FileInfo`?

File: server/status.go (hunk 23 above)
Comment (BramGruneir) [inline server/status.go]: fixed

File: util/log/file.go (hunk 21 above)
Comment (BramGruneir) [inline util/log/file.go]: fixed

File: util/log/file.go (hunk 24 above)
Comment (tamird) [inline util/log/file.go]: Sorry, I don't understand. Are you saying the error string is inspected programmatically somewhere?

File: util/log/file.go (hunk 22 above)
Comment (BramGruneir) [inline util/log/file.go]: of course, done.

File: util/log/file.go (hunk 24 above)
Comment (BramGruneir) [inline util/log/file.go]: oh, I see what you mean, sure.

File: util/log/file.go (hunk 25 above)
Comment (BramGruneir) [inline util/log/file.go]: sure.  done.

File: util/log/file.go (hunk 26 above)
Comment (BramGruneir) [inline util/log/file.go]: makes sense, done.

File: util/log/file.go [hunk 30]
```diff
@@ -85,36 +97,92 @@ func shortHostname(hostname string) string {
...
+// escapeString replaces all the periods in the string with an underscore, and
+// every underscore with a double underscore.
+func escapeStringForFilename(s string) string {
+	sEscapedPartial := strings.Replace(s, "_", "__", -1)
+	return strings.Replace(sEscapedPartial, ".", "_", -1)
+}
+
+// unescapeStringForFilename reverts the escaping from escapeStringForFilename.
+func unescapeStringForFilename(s string) string {
+	sUnescapedPartial := strings.Replace(s, "_", ".", -1)
```
Comment (tschottdorf) [inline util/log/file.go]: when this is done, there won't be any underscores left, so the next replace is always a no-op, isn't it?

File: util/log/file.go [hunk 31]
```diff
@@ -85,36 +97,92 @@ func shortHostname(hostname string) string {
...
-		t.Minute(),
-		t.Second(),
+// escapeString replaces all the periods in the string with an underscore, and
+// every underscore with a double underscore.
+func escapeStringForFilename(s string) string {
+	sEscapedPartial := strings.Replace(s, "_", "__", -1)
+	return strings.Replace(sEscapedPartial, ".", "_", -1)
+}
+
+// unescapeStringForFilename reverts the escaping from escapeStringForFilename.
```
Comment (tschottdorf) [inline util/log/file.go]: this needs to be qualified more because the escaping maps a lot of stuff to the same string, for instance `..` and `_` both end up as `__`.

File: util/log/file.go [hunk 32]
```diff
@@ -85,36 +97,92 @@ func shortHostname(hostname string) string {
...
-		host,
-		userName,
-		tag,
-		t.Year(),
-		t.Month(),
-		t.Day(),
-		t.Hour(),
-		t.Minute(),
-		t.Second(),
+// escapeString replaces all the periods in the string with an underscore, and
```
Comment (tschottdorf) [inline util/log/file.go]: the order in which this happens is opposite to what is suggested here, otherwise `..` would encode to `____` but it ends up being `__`.

File: util/log/file.go (hunk 22 above)
Comment (tschottdorf) [inline util/log/file.go]: I'm assuming this is only used for a fixed format, so maybe easier to just deal with that format rather than trying to have something general which needs to think about all sorts of situations irrelevant to you.

File: util/log/clog.go (hunk 27 above)
Comment (BramGruneir) [inline util/log/clog.go]: I think you're right, it doesn't need to be exported.  I unexported it.

And doesn't stringer add .String(), which is already used by Severity to display the number?

Comment (tamird) [inline util/log/clog.go]: i'm actually not sure how stringer would help here since these are not exactly 1:1

Comment (tschottdorf) [inline util/log/clog.go]: yep, I thought this translation here was used for String(), which isn't the case. Carry on.

File: util/log/file.go (hunk 29 above)
Comment (BramGruneir) [inline util/log/file.go]: Brilliant idea.  Reduced the size of the function and it's one lass pass through the list of files.

File: util/log/file.go (hunk 28 above)
Comment (BramGruneir) [inline util/log/file.go]: Renamed to getFileDetails

File: util/log/file.go (hunk 30 above)
Comment (BramGruneir) [inline util/log/file.go]: Ah, you're right.  I think I'm just going to remove all periods from the filename.  And not try to restore them.

File: util/log/file.go (hunk 22 above)
Comment (BramGruneir) [inline util/log/file.go]: I'm just going to strip the extra periods that don't mesh with our formatting.

File: util/log/file.go (hunk 31 above)
Comment (BramGruneir) [inline util/log/file.go]: I replaced this by removing all periods from the strings instead of escaping them.

File: util/log/file.go (hunk 32 above)
Comment (BramGruneir) [inline util/log/file.go]: Yeah.  Was all kinds of broken.  Removed.

**Comment by BramGruneir** [comment]
All comments addressed, PTAL.
If it's all good, I'll squish it up before merging.

File: server/status.go [hunk 33]
```diff
@@ -223,6 +240,99 @@ func (s *statusServer) handleLocalLog(w http.ResponseWriter, r *http.Request, ps
...
+	level := ps.ByName("level")
+	var sev log.Severity
+	if len(level) == 0 {
+		sev = log.InfoLog
+	} else {
+		var sevFound bool
+		sev, sevFound = log.SeverityByName(level)
+		if !sevFound {
+			w.WriteHeader(http.StatusBadRequest)
+			w.Write([]byte(fmt.Sprintf("could not determine level %s", level)))
```
Comment (tamird) [inline server/status.go]: I don't know if it will work, but http://golang.org/pkg/io/#WriteString will save you an allocation

Comment (tamird) [inline server/status.go]: dep, just use `fmt.Fprintf`

Comment (BramGruneir) [inline server/status.go]: derp indeed!  Fixed.

File: server/status_test.go [hunk 34]
```diff
@@ -337,20 +346,89 @@ func TestStatusLocalLog(t *testing.T) {
...
+		{log.InfoLog, 0, timestamp, timestampE, true, false, false},
+		{log.InfoLog, 0, timestampE, timestampEW, false, true, false},
+		{log.InfoLog, 0, timestampEW, timestampEWI, false, false, true},
+		{log.InfoLog, 0, timestamp, timestampEW, true, true, false},
+		{log.InfoLog, 0, timestampE, timestampEWI, false, true, true},
+		{log.InfoLog, 0, timestamp, timestampEWI, true, true, true},
+	}
+
+	for i, testCase := range testCases {
+		url := fmt.Sprintf("%s%s?", statusLocalLogKeyPrefix, testCase.Level.Name())
```
Comment (tamird) [inline server/status_test.go]: nit: allocating tons of strings here, which can be avoided with Fprintf

File: util/log/file.go [hunk 35]
```diff
@@ -85,36 +97,86 @@ func shortHostname(hostname string) string {
...
+func removePeriods(s string) string {
+	return strings.Replace(s, ".", "", -1)
+}
+
+// logName returns a new log file name containing the severity, with start time
+// t, and the name for the symlink for the severity.
+func logName(severity Severity, t time.Time) (name, link string) {
+	// Replace the ':'s in the time format with '_'s to allow for log files in
+	// Windows.
+	tFormatted := strings.Replace(t.Format(time.RFC3339), ":", "_", -1)
```
Comment (tamird) [inline util/log/file.go]: I think I would extract this into a method (which can have a nice descriptive name like `escapeForWindows`)

File: util/log/file.go [hunk 36]
```diff
@@ -85,36 +97,86 @@ func shortHostname(hostname string) string {
...
+		return FileDetails{}, util.Errorf("malformed log filename")
+	}
+
+	sev, sevFound := SeverityByName(matches[4])
+	if !sevFound {
+		return FileDetails{}, util.Errorf("malformed severity")
+	}
+
+	// Replace the '_'s with ':'s to restore the correct time format.
+	fixTime := strings.Replace(matches[5], "_", ":", -1)
```
Comment (tamird) [inline util/log/file.go]: ditto re extraction

File: util/log/file.go [hunk 37]
```diff
@@ -221,3 +291,101 @@ func GetLogReader(filename string, allowAbsolute bool) (io.ReadCloser, error) {
 	}
 	return nil, err
 }
+
+// int64sortable is required so we can sort on an int64 slice.
+type fileInfosortable []FileInfo
```
Comment (tamird) [inline util/log/file.go]: :star2:

Comment (tamird) [inline util/log/file.go]: even more greppable would be sortableFileInfo

File: util/log/file.go [hunk 38]
```diff
@@ -221,3 +291,101 @@ func GetLogReader(filename string, allowAbsolute bool) (io.ReadCloser, error) {
...
+func (a fileInfosortable) Len() int      { return len(a) }
+func (a fileInfosortable) Swap(i, j int) { a[i], a[j] = a[j], a[i] }
+func (a fileInfosortable) Less(i, j int) bool {
+	return a[i].Details.Time.UnixNano() < a[j].Details.Time.UnixNano()
+}
+
+// selectFiles selects all log files that have an timestamp before the endTime and
+// the correct severity. It then sorts them in decreasing order, with the most
+// recent as the first one.
+func selectFiles(logFiles []FileInfo, severity Severity, endTimestamp int64) []FileInfo {
```
Comment (tamird) [inline util/log/file.go]: `filesBeforeTimeDescending`? something to indicate that this sorts, too

**Comment by tamird** [comment]
Some final nits then LGTM

File: util/log/file.go (hunk 35 above)
Comment (BramGruneir) [inline util/log/file.go]: It's only used in one place.  I don't think it requires it's own function.

File: util/log/file.go (hunk 36 above)
Comment (BramGruneir) [inline util/log/file.go]: again, only used here.  If it is needed anywhere else, I would do so.

File: util/log/file.go (hunk 37 above)
Comment (BramGruneir) [inline util/log/file.go]: done!

File: util/log/file.go (hunk 36 above)
Comment (tamird) [inline util/log/file.go]: i think the function would clarify the intent, but again, i don't feel strongly

File: util/log/file.go (hunk 38 above)
Comment (BramGruneir) [inline util/log/file.go]: That just sounds too verbose.  In most other modern languages, I could do this whole thing in a single lambda expression (or a collection of chained calls).  

logFiles.where(z => z.Details.Severity == severity).where(z => z.Details.Time.UnixNano() < endTimestamp).orderBy(z => z.Details.Time.UnixNano())

It's a tiny function and it's clear from the documentation that it sorts as well.  A select statement usually has some sorting in it.

Comment (tamird) [inline util/log/file.go]: > A select statement usually has some sorting in it.

Totally disagree with this, but leave the name as-is if you prefer.

File: server/status_test.go (hunk 34 above)
Comment (BramGruneir) [inline server/status_test.go]: switched it to a bytes.Buffer and Fprintf

**Comment by BramGruneir** [comment]
Thanks for the reviews everyone.  Going to merge this on green.

File: server/status.go [hunk 39]
```diff
@@ -51,13 +54,22 @@ const (
 	statusLocalKeyPrefix = statusKeyPrefix + "local/"
 
 	// statusLocalLogKeyPrefix exposes a list of log files for the node.
```
Comment (spencerkimball) [inline server/status.go]: s/statusLocalLogKeyPrefix/statusLocalLogFileKeyPrefix/

**Comment by spencerkimball** [comment]
Wow big change. This is great.
